pkg/types/contract_reader_cache.go

```diff
@@ -0,0 +1,441 @@
+package types
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"reflect"
+	"sync"
+	"time"
+
+	"github.com/smartcontractkit/chainlink-common/pkg/services"
+	"github.com/smartcontractkit/chainlink-common/pkg/types/query"
+	"github.com/smartcontractkit/chainlink-common/pkg/types/query/primitives"
+	"github.com/smartcontractkit/chainlink-common/pkg/utils"
+)
+
+var (
+	ErrNotExist        = errors.New("data does not exist for key")
+	ErrTypeMismatch    = errors.New("data type mismatch")
+	ErrInvalidStrategy = errors.New("invalid cache strategy")
+)
+
+const (
+	CachedContractReaderName = "CachedContractReader"
+
+	cleanInterval = time.Second
+	pollInterval  = time.Second
+)
+
+type TriggerAndCache struct {
+	Params any
+	Value  any
+}
+
+type PollAndCache struct {
+	TriggerAndCache
+	Interval time.Duration
+}
+
+type FallthroughOnStale struct {
+	Staleness time.Duration
```

Comment:
I wonder if we can find a more intuitive (and more explicit?) name for this type and field. Maybe with `Expiration`/`Expires`?
Google Java caches offer `expireAfterAccess` and `expireAfterWrite`: https://github.com/google/guava/wiki/CachesExplained#timed-eviction

Also, we should consider offering this option _alongside_ the other strategies, because consider e.g. if your RPCs are unavailable for extended period of time:
- triggers are being missed
- polling is failing

but the cached value is still being reported without any errors, which is often incorrect though can be subtle and overlooked.

Fortunately, I think this could be a simplifying change overall, if we reduce the different strategy types to a single struct:
```go
type CacheStrategy struct {
	Params any
	Value any

	PollInterval time.Duration

	ExpiresAfter time.Duration
}
```
Then we can infer nocache, trigger, and poll based on which fields are set, while applying expiration to all.